reactos/drivers/usb/usbport/iface.c

```diff
@@ -0,0 +1,916 @@
+    PdoExtension->RootHubInitCallback = CallbackFunction;
+    KeReleaseSpinLock(&FdoExtension->RootHubCallbackSpinLock, OldIrql);
+
+    return STATUS_SUCCESS;
+}
+
+VOID
+USB_BUSIFFN
+USBHI_FlushTransfers(IN PVOID BusContext,
+                     OUT PUSB_DEVICE_HANDLE UsbdDeviceHandle)
+{
+    PDEVICE_OBJECT PdoDevice;
+    PUSBPORT_RHDEVICE_EXTENSION PdoExtension;
+
+    DPRINT("USBHI_FlushTransfers: ... \n");
+
+    PdoDevice = (PDEVICE_OBJECT)BusContext;
+    PdoExtension = (PUSBPORT_RHDEVICE_EXTENSION)PdoDevice->DeviceExtension;
+
+    USBPORT_BadRequestFlush(PdoExtension->FdoDevice);
+}
+
+VOID
+USB_BUSIFFN
+USBHI_SetDeviceHandleData(IN PVOID BusContext,
+                          IN PVOID DeviceHandle,
+                          IN PDEVICE_OBJECT UsbDevicePdo)
+{
+    DPRINT1("USBHI_SetDeviceHandleData: UNIMPLEMENTED. FIXME. \n");
+}
+
+/* USB bus driver Interface functions */
+
+VOID
+USB_BUSIFFN
+USBDI_GetUSBDIVersion(IN PVOID BusContext,
+                      OUT PUSBD_VERSION_INFORMATION VersionInfo,
+                      OUT PULONG HcdCapabilities)
+{
+    DPRINT1("USBDI_GetUSBDIVersion: UNIMPLEMENTED. FIXME. \n");
+}
+
+NTSTATUS
+USB_BUSIFFN
+USBDI_QueryBusTime(IN PVOID BusContext,
+                   OUT PULONG CurrentFrame)
+{
+    DPRINT1("USBDI_QueryBusTime: UNIMPLEMENTED. FIXME. \n");
+    return STATUS_SUCCESS;
+}
+
+NTSTATUS
+USB_BUSIFFN
+USBDI_SubmitIsoOutUrb(IN PVOID BusContext,
+                      IN PURB Urb)
+{
+    DPRINT1("USBDI_SubmitIsoOutUrb: UNIMPLEMENTED. FIXME. \n");
+    return STATUS_SUCCESS;
+}
+
+NTSTATUS
+USB_BUSIFFN
+USBDI_QueryBusInformation(IN PVOID BusContext,
+                          IN ULONG Level,
+                          OUT PVOID BusInfoBuffer,
+                          OUT PULONG BusInfoBufferLen,
+                          OUT PULONG BusInfoActualLen)
+{
+    PDEVICE_OBJECT PdoDevice;
+    PUSBPORT_RHDEVICE_EXTENSION PdoExtension;
+    PDEVICE_OBJECT FdoDevice;
+    PUSBPORT_DEVICE_EXTENSION FdoExtension;
+    SIZE_T Length;
+    //PUSB_BUS_INFORMATION_LEVEL_0 Buffer0;
+    PUSB_BUS_INFORMATION_LEVEL_1 Buffer1;
+
+    DPRINT("USBDI_QueryBusInformation: Level - %p\n", Level);
+
+    if ((Level != 0) || (Level != 1))
+    {
+        DPRINT1("USBDI_QueryBusInformation: Level should be 0 or 1\n");
+        return STATUS_NOT_SUPPORTED;
+    }
+
+    PdoDevice = (PDEVICE_OBJECT)BusContext;
+    PdoExtension = (PUSBPORT_RHDEVICE_EXTENSION)PdoDevice->DeviceExtension;
+    FdoDevice = PdoExtension->FdoDevice;
+    FdoExtension = (PUSBPORT_DEVICE_EXTENSION)FdoDevice->DeviceExtension;
+
+    if (Level == 0)
+    {
+        if (BusInfoActualLen)
+            *BusInfoActualLen = sizeof(USB_BUS_INFORMATION_LEVEL_0);
+    
+        if (*BusInfoBufferLen < sizeof(USB_BUS_INFORMATION_LEVEL_0))
+        {
+            return STATUS_BUFFER_TOO_SMALL;
+        }
+    
+        *BusInfoBufferLen = sizeof(USB_BUS_INFORMATION_LEVEL_0);
+    
+        //Buffer0 = (PUSB_BUS_INFORMATION_LEVEL_0)BusInfoBuffer;
```

Comment:
In accordance with https://msdn.microsoft.com/en-us/library/windows/hardware/ff537620(v=vs.85).aspx it func. should return:
TotalBandwidth - Specifies the total bandwidth, in bits per second, available on the bus:
ConsumedBandwidth - Specifies the mean bandwidth that is already in use, in bits per second;
Also STATUS_SUCCESS or STATUS_BUFFER_TOO_SMALL.
Than this function is really used I do not have information yet.
Any suggestions?
In general, the use of the bandwidth of the USB bus is a large amount of information in the USB specification (Chapter 5 USB Data Flow Model). 
reactos\sdk\lib\drivers\libusb\hub_controller.cpp only return STATUS_NOT_IMPLEMENTED.